### PR TITLE
Silent Payments: receiving

### PR DESCRIPTION
**Pull request by josibake**
This PR is a child of https://github.com/bitcoin/bitcoin/pull/27827 and only implements the receiving logic. For the rest:

* #28122 - implement the silent payments scheme
* #28201 - implements sending
* #27827 - all three PRs together - this PR is meant for tracking progress and is useful for reviewers who want to compile all three PRs at once for testing

This PR depends on #28122 and is marked as a draft until it is merged. Commits up to https://github.com/bitcoin/bitcoin/commit/8ee791e7b34a354ce6835b968fe12924b8908d7c belong to #28122; please review those commits on #28122

## Receiving 

_Description coming soon_

### CODE REVIEW

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -1449,6 +1574,134 @@ void CWallet::transactionRemovedFromMempool(const CTransactionRef& tx, MemPoolRe
...
+
+                    // Get the public key from the back of the stack
+                    // No need to check if it actually hashes to the hash in the scriptPubKey as
+                    // this has already been checked
+                    CPubKey pubkey{stack.back()};
+                    assert(pubkey.IsFullyValid());
+                    pubkeys_for_shared_secret.push_back(pubkey);
+                    continue;
+                }
+            case TxoutType::SCRIPTHASH:
```
Comment (achow101) [inline src/wallet/wallet.cpp:1676]: Should recurse on `SCRIPTHASH` instead of assuming.

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -1852,6 +2128,24 @@ int64_t CWallet::RescanFromTime(int64_t startTime, const WalletRescanReserver& r
     return startTime;
 }
 
+Coin CWallet::FindPreviousCoin(const CTxIn& txin) const
+{
+    // Find the UTXO being spent in UTXO Set to learn the transaction type
+    std::map<COutPoint, Coin> coins;
+    coins[txin.prevout];
+    chain().findCoins(coins);
```
Comment (achow101) [inline src/wallet/wallet.cpp:2136]: This looks in the UTXO set but is called on a transaction that has already been verified and so therefore it's inputs are no longer in the UTXO set. This results in the `Coin` being uninitialized which causes later `Solver` calls to fail unexpectedly.

Comment (achow101) [reply src/wallet/wallet.cpp:2136]: As discussed offline, we shouldn't use any view of the UTXO set. Rather we should use the undo data when receiving a `blockConnected`, and modify `TransactionAddedToMempool` to include the previous outputs. This lets us avoid any race conditions with the state of the UTXO set and guarantees that the previous output data will always be available for the silent payments calculations.

**Comment by BrandonOdiwuor** [comment]
Tested ACK with some comments at the end

Was able to successfully run unit and functional tests on 'pr/28202' branch and  verified that the tests fail against the master branch

Manually tested on regtest as described below:
```sh
$ ./src/bitcoin-cli -regtest -named createwallet wallet_name="sp-wallet" silent_payment=true
```
```sh
$ ./src/bitcoin-cli -regtest  getwalletinfo # showed that a wallet with silent payment enabled was created successfully

{
    "walletname": "sp-wallet",
    "walletversion": 169900,
    "format": "sqlite",
    "balance": 0.00000000,
    "unconfirmed_balance": 0.00000000,
    "immature_balance": 0.00000000,
    "txcount": 0,
    "keypoolsize": 4001,
    "keypoolsize_hd_internal": 4000,
    "paytxfee": 0.00000000,
    "private_keys_enabled": true,
    "avoid_reuse": false,
    "scanning": false,
    "descriptors": true,
    "external_signer": false,
    "blank": false,
    "silent_payment": true,
    "lastprocessedblock": {
        "hash": "2fbd2f41ecd9e553cc8d129cff71846e0f26da7584eb8362e7cc70c0913b8e98",
        "height": 212
    }
}
```

```sh
$ ./src/bitcoin-cli -regtest listdescriptors # was able to list a silent payment descriptor
{
…
    {                                                                                                                                                                                                              
        "desc": "sp([f3a37e3d/352h/1h/0h/1h]tpubDFMFwrp8wbV1voiEY3dy6ba5mv3nzbzRQmBHYfu6c3gGnf49QM9PB6Uy4P3Rpb5YcgXHT4vcXCpWiDt8rSoQXceJenU2WkvxcV6rYAxVxm6/0,[f3a37e3d/352h/1h/0h/0h]tpubDFMFwrp8wbV1tkGrdjqb8VSRp7srJQGvtqxmBSU3LVySKFPzCH1sP9cuLtUHA2Cz6qFSXdi2F52gAGNzqMrZrGrNwmf7EzbxebVEzBKugvm/0)#qzhxnezt",
        "timestamp": 1692181466,
        "active": true,
        "internal": false,
        "next": 0
    },
…
}
```
```sh
$ ./src/bitcoin-cli -regtest  getsilentaddress # was able to generate a silent payment address
{
    "address": "sprt1qq0yr8yv2dzrrvkyde8lex6qjy2jklgpxud260yvepgdfh5rh2ldj2qket9cwqj9xt5j7mseqk4appaqzung4xhwknaxcchlf54vutvv4mcper7ca"
}
```
```sh
$ ./src/bitcoin-cli -regtest  decodesilentaddress sprt1qq0yr8yv2dzrrvkyde8lex6qjy2jklgpxud260yvepgdfh5rh2ldj2qket9cwqj9xt5j7mseqk4appaqzung4xhwknaxcchlf54vutvv4mcper7ca 
# was able to correctly decode the silent payment address above into scan and spen pubkeys
{
    "address": "sprt1qq0yr8yv2dzrrvkyde8lex6qjy2jklgpxud260yvepgdfh5rh2ldj2qket9cwqj9xt5j7mseqk4appaqzung4xhwknaxcchlf54vutvv4mcper7ca",
    "scan_pubkey": "03c833918a688636588dc9ff93681222a56fa026e355a791990a1a9bd07757db25",
    "spend_pubkey": "02d95970e048a65d25edc320b57a10f402e4d1535dd69f4d8c5fe9a559c5b195de"
}
```

## comments

- I noticed that ```./src/bitcoin-cli -regtest listdescriptors``` only listed one descriptor for silent payments
- Also that ```./src/bitcoin-cli -regtest  getsilentaddress```  always return the same one address every time

**Review by BrandonOdiwuor — COMMENTED** [review]
Tested ACK with some comments at the end

Was able to successfully run unit and functional tests on 'pr/28202' branch and  verified that the tests fail against the master branch

Manually tested on regtest as described below:
```sh
$ ./src/bitcoin-cli -regtest -named createwallet wallet_name="sp-wallet" silent_payment=true
```
```sh
$ ./src/bitcoin-cli -regtest  getwalletinfo # showed that a wallet with silent payment enabled was created successfully

{
    "walletname": "sp-wallet",
    "walletversion": 169900,
    "format": "sqlite",
    "balance": 0.00000000,
    "unconfirmed_balance": 0.00000000,
    "immature_balance": 0.00000000,
    "txcount": 0,
    "keypoolsize": 4001,
    "keypoolsize_hd_internal": 4000,
    "paytxfee": 0.00000000,
    "private_keys_enabled": true,
    "avoid_reuse": false,
    "scanning": false,
    "descriptors": true,
    "external_signer": false,
    "blank": false,
    "silent_payment": true,
    "lastprocessedblock": {
        "hash": "2fbd2f41ecd9e553cc8d129cff71846e0f26da7584eb8362e7cc70c0913b8e98",
        "height": 212
    }
}
```

```sh
$ ./src/bitcoin-cli -regtest listdescriptors # was able to list a silent payment descriptor
{
…
    {                                                                                                                                                                                                              
        "desc": "sp([f3a37e3d/352h/1h/0h/1h]tpubDFMFwrp8wbV1voiEY3dy6ba5mv3nzbzRQmBHYfu6c3gGnf49QM9PB6Uy4P3Rpb5YcgXHT4vcXCpWiDt8rSoQXceJenU2WkvxcV6rYAxVxm6/0,[f3a37e3d/352h/1h/0h/0h]tpubDFMFwrp8wbV1tkGrdjqb8VSRp7srJQGvtqxmBSU3LVySKFPzCH1sP9cuLtUHA2Cz6qFSXdi2F52gAGNzqMrZrGrNwmf7EzbxebVEzBKugvm/0)#qzhxnezt",
        "timestamp": 1692181466,
        "active": true,
        "internal": false,
        "next": 0
    },
…
}
```
```sh
$ ./src/bitcoin-cli -regtest  getsilentaddress # was able to generate a silent payment address
{
    "address": "sprt1qq0yr8yv2dzrrvkyde8lex6qjy2jklgpxud260yvepgdfh5rh2ldj2qket9cwqj9xt5j7mseqk4appaqzung4xhwknaxcchlf54vutvv4mcper7ca"
}
```
```sh
$ ./src/bitcoin-cli -regtest  decodesilentaddress sprt1qq0yr8yv2dzrrvkyde8lex6qjy2jklgpxud260yvepgdfh5rh2ldj2qket9cwqj9xt5j7mseqk4appaqzung4xhwknaxcchlf54vutvv4mcper7ca 
# was able to correctly decode the silent payment address above into scan and spen pubkeys
{
    "address": "sprt1qq0yr8yv2dzrrvkyde8lex6qjy2jklgpxud260yvepgdfh5rh2ldj2qket9cwqj9xt5j7mseqk4appaqzung4xhwknaxcchlf54vutvv4mcper7ca",
    "scan_pubkey": "03c833918a688636588dc9ff93681222a56fa026e355a791990a1a9bd07757db25",
    "spend_pubkey": "02d95970e048a65d25edc320b57a10f402e4d1535dd69f4d8c5fe9a559c5b195de"
}
```

## comments

- I noticed that ```./src/bitcoin-cli -regtest listdescriptors``` only listed one descriptor for silent payments
- Also that ```./src/bitcoin-cli -regtest  getsilentaddress```  always return the same one address every time

**Comment by josibake** [comment]
Closing in favor of https://github.com/bitcoin/bitcoin/pull/28453

Originally, the idea was to implement sending and receiving separately, but it makes more sense to have them together so the receiving PR can use silent payments for generating change.

**Pull request closed without merging**